Make Land activity self-contained and allow explicit landing orders

tovl opened the pull request:
Depends on #16417 and #16365.

Implements features requested in https://github.com/OpenRA/OpenRA/pull/16417#pullrequestreview-229217672.
Implements refactor requested in https://github.com/OpenRA/OpenRA/pull/16365#discussion_r279190928.
Fixes #16331.
Fixes #12464.
Fixes #12201.
Fixes #4642.
Fixes #12081.
Fixes #11748.
Fixes #16572.
Fixes #16646.

This PR moves all the approach before landing logic into the `Land` activity itself. This includes flying and turning for VTOL aircraft. It also includes and generalizes the approach logic for non-VTOL aircraft found in `ReturnToBase` and the logic from `DeliverUnit` that looks for suitable landing locations. This makes the `Land` activity completely self-contained and flexible. No other activity should bother with the specifics of landing, but just queue a `Land` activity with a desired `target,` `offset` and `facing` (all optional parameters) instead.

Furthermore, this implements explicit land orders and reworks the carryall and transport logic to support transport craft remaining airborne when idle.

Changes in behaviour:
- Orca carryalls, orca transports and chinooks remain airborn when idle.
- All aircraft with `LandableTerraintypes` defined can be ordered to land at a location using force-move.
- All VTOL aircraft in the default mods have `LandableTerrainTypes` defined and can therefore be ordered to land.
- When ordered to land at a cell (not a docking actor) occupied by an actor that cannot be crushed or nudged, aircraft will land at a nearby cell instead. (This was generalized from `DeliverUnit`)
- When infantry try to enter an aerial transport, it will automatically land and the infantry will wait for it to do so.
- Carryalls will automatically lift off after picking up or delivering a unit.
- Scripted unload orders may optionally include a destination and `UnloadCargo` will then take care of getting close to that destination.
- Non-VTOL carryalls are fully supported. (Replace `Inherits: ^Helicopter` with `Inherits: ^Aircraft` in the TS carryall to test)

pchote commented:
This will be a very nice way to cap off the aircraft changes for the playtest, so adding to the milestone.

pchote commented:
I few issues I noticed while quickly testing:
* Please add "- Helicopters will land at the target location" to the Force Move command bar button tooltip in RA/TD/TS.
* Force move no longer works at all for Yak/Mig - they should fall back to a regular move instead.
* IMO it feels inconsistent to have transport helicopters stay on the ground after anything enters them. By default they should take off again once all reserved passengers have entered them, and after unloading - but care will be needed to not break scripted helicopters in missions.
* The RA shellmap breaks with a lua error:
   ```
   Fatal Lua Error: [string "BindingSupport.lua"]:30: Argument 'Nullable`1 destination' of 'void UnloadPassengers(Nullable`1 destination)' is not optional.
     at OpenRA.Scripting.ScriptContext.FatalError (System.String message) [0x00000] in <1794848e7ea34e06a09d113c2a6c7693>:0 
     at OpenRA.Mods.Common.Scripting.ScriptTriggers.OpenRA.Mods.Common.Traits.INotifyDamage.Damaged (OpenRA.Actor self, OpenRA.Traits.AttackInfo e) [0x00069] in <501f99702a4548b3b3aace44ed353658>:0 
     at OpenRA.Mods.Common.Traits.Health.InflictDamage (OpenRA.Actor self, OpenRA.Actor attacker, OpenRA.Traits.Damage damage, System.Boolean ignoreModifiers) [0x000e9] in <501f99702a4548b3b3aace44ed353658>:0 
     at OpenRA.Actor.InflictDamage (OpenRA.Actor attacker, OpenRA.Traits.Damage damage) [0x00011] in <1794848e7ea34e06a09d113c2a6c7693>:0 
     at OpenRA.Mods.Common.Warheads.SpreadDamageWarhead.DoImpact (OpenRA.WPos pos, OpenRA.Actor firedBy, System.Collections.Generic.IEnumerable`1[T] damageModifiers) [0x00160] in <501f99702a4548b3b3aace44ed353658>:0 
     at OpenRA.Mods.Common.Warheads.DamageWarhead.DoImpact (OpenRA.Traits.Target target, OpenRA.Actor firedBy, System.Collections.Generic.IEnumerable`1[T] damageModifiers) [0x000b1] in <501f99702a4548b3b3aace44ed353658>:0 
     at OpenRA.GameRules.WeaponInfo.Impact (OpenRA.Traits.Target target, OpenRA.Actor firedBy, System.Collections.Generic.IEnumerable`1[T] damageModifiers) [0x00076] in <1794848e7ea34e06a09d113c2a6c7693>:0 
     at OpenRA.Mods.Common.Traits.Explodes.OpenRA.Mods.Common.Traits.INotifyKilled.Killed (OpenRA.Actor self, OpenRA.Traits.AttackInfo e) [0x00138] in <501f99702a4548b3b3aace44ed353658>:0 
     at OpenRA.Mods.Common.Traits.Health.InflictDamage (OpenRA.Actor self, OpenRA.Actor attacker, OpenRA.Traits.Damage damage, System.Boolean ignoreModifiers) [0x00249] in <501f99702a4548b3b3aace44ed353658>:0 
     at OpenRA.Actor.InflictDamage (OpenRA.Actor attacker, OpenRA.Traits.Damage damage) [0x00011] in <1794848e7ea34e06a09d113c2a6c7693>:0 
     at OpenRA.Mods.Common.Warheads.SpreadDamageWarhead.DoImpact (OpenRA.WPos pos, OpenRA.Actor firedBy, System.Collections.Generic.IEnumerable`1[T] damageModifiers) [0x00160] in <501f99702a4548b3b3aace44ed353658>:0 
     at OpenRA.Mods.Common.Warheads.DamageWarhead.DoImpact (OpenRA.Traits.Target target, OpenRA.Actor firedBy, System.Collections.Generic.IEnumerable`1[T] damageModifiers) [0x000b1] in <501f99702a4548b3b3aace44ed353658>:0 
     at OpenRA.GameRules.WeaponInfo.Impact (OpenRA.Traits.Target target, OpenRA.Actor firedBy, System.Collections.Generic.IEnumerable`1[T] damageModifiers) [0x00076] in <1794848e7ea34e06a09d113c2a6c7693>:0 
     at OpenRA.Mods.Common.Projectiles.Bullet.Explode (OpenRA.World world) [0x00032] in <501f99702a4548b3b3aace44ed353658>:0 
     at OpenRA.Mods.Common.Projectiles.Bullet.Tick (OpenRA.World world) [0x002e7] in <501f99702a4548b3b3aace44ed353658>:0 
     at OpenRA.World.<Tick>b__103_1 (OpenRA.Effects.IEffect e) [0x00000] in <1794848e7ea34e06a09d113c2a6c7693>:0 
     at OpenRA.WorldUtils.DoTimed[T] (System.Collections.Generic.IEnumerable`1[T] e, System.Action`1[T] a, System.String text) [0x00015] in <1794848e7ea34e06a09d113c2a6c7693>:0 
     at OpenRA.World.Tick () [0x00180] in <1794848e7ea34e06a09d113c2a6c7693>:0 
     at OpenRA.Game.InnerLogicTick (OpenRA.Network.OrderManager orderManager) [0x00206] in <1794848e7ea34e06a09d113c2a6c7693>:0 
     at OpenRA.Game.LogicTick () [0x0004a] in <1794848e7ea34e06a09d113c2a6c7693>:0 
     at OpenRA.Game.Loop () [0x000e0] in <1794848e7ea34e06a09d113c2a6c7693>:0 
     at OpenRA.Game.Run () [0x0003c] in <1794848e7ea34e06a09d113c2a6c7693>:0 
     at OpenRA.Game.InitializeAndRun (System.String[] args) [0x00010] in <1794848e7ea34e06a09d113c2a6c7693>:0 
     at OpenRA.Program.Main (System.String[] args) [0x00044] in <1794848e7ea34e06a09d113c2a6c7693>:0 
   ```

pchote commented on OpenRA.Mods.Common/Scripting/Properties/TransportProperties.cs:
```diff
@@ -42,9 +42,9 @@ public TransportProperties(ScriptContext context, Actor self)
 		[ScriptActorPropertyActivity]
 		[Desc("Command transport to unload passengers.")]
-		public void UnloadPassengers()
+		public void UnloadPassengers(Target? destination)
```
Setting a default null value should be enough to fix the API breakage.

tovl replied:
OK, that's a silly mistake. The whole point of making `destination` nullable was to avoid breaking the API, but then I forgot to actually add a default. :man_facepalming:

tovl commented:
Applied all fixups.

PunkPun reviewed (commented):
So far I really enjoy the changes. Bugs:

- When the chinook is ordered to land with alt-move infantry scatter away from the landing spot. When infantry order the chinook to land, it doesn't do that and just hovers in air. 
- Alt-load doesn't work.
- If a chinook in air is sharing a cell with another aircraft it will not land when infantry order it to

These bugs most likely also apply to #16417

tovl commented:
> When the chinook is ordered to land with alt-move infantry scatter away from the landing spot. When infantry order the chinook to land, it doesn't do that and just hovers in air.

Since the infantry is not really idle in that case, they cannot be nudged by the chinook. I've made the `Land` activity consider non-idle allied units blocking so they will now land adjacent to them instead.

> Alt-load doesn't work.

Fixed.

> If a chinook in air is sharing a cell with another aircraft it will not land when infantry order it to

This is actually just one expression of a very nasty bug in the aircraft repulsion logic. Aircraft can only be repulsed by other aircraft that were build after them. They will not budge otherwise, no matter which one is moving. This is probably the underlying cause of most other issues with aircraft blocking each-other. Fixing this is beyond the scope of this PR (and I haven't been successful at finding the exact cause anyway).

PunkPun reviewed (commented):
It feels much better, the mid cell thingie is really annoying though

- If a chinook is issued a stop command when landing/ascending it will hover in that height (all helicopters do this)
- If a chinook is ordered to land by infantry and then immediately the order is cancelled the chinook will stay landed
- I think the alt-load distance needs to be increased, it's low for such big units

reaperrr commented:
> If a chinook is issued a stop command when landing/ascending it will hover in that height (all helicopters do this)

I think that 'Land' should either be non-interruptible, or aircraft should rise back to CruiseAltitude if given a Stop command while not on ground.

tovl commented:
>  think that 'Land' should either be non-interruptible, or aircraft should rise back to CruiseAltitude if given a Stop command while not on ground.

`Land` is now uninterruptible from the moment the landing cell is reserved.

> If a chinook is ordered to land by infantry and then immediately the order is cancelled the chinook will stay landed

Fixed.

> I think the alt-load distance needs to be increased, it's low for such big units

This is a matter of preference, but I tend to agree. The way I see it, if you need more control over which transports the units go to you can always use shift click. I've doubled the default range.

PunkPun reviewed (changes requested):
- Landing now is uninterruptible but it shouldn't be the case, if an order of any kind is given that doesn't require it to continue landing it should come back to the cruise altitude and do the action 
- Ascending can still be stopped midway

Even though the range of alt-load was increased it seems as it is the same as it was previously, I guess the logic behind it has to be reworked. Also I never considered queueing up loading but I guess since all behaviours are now queueable I should start thinking with them.

tovl commented:
Helis that are not on the ground and do not have the `LandOnIdle` flag will now always rise to `CruiseAltitude` when idle.

I've found that the alt-load behaviour isn't even consistent so this indicates a bigger problem. This isn't directly related to this PR so I'll defer a fix to future PRs.

PunkPun reviewed (changes requested):
Works almost perfectly, the last complaint I have is when I select a chinook that is over ground based units the units below are going to be prioritised

tovl commented:
This is true for all aircraft and is pretty out of scope for this PR.

pchote commented on OpenRA.Mods.Common/Traits/Carryall.cs:
```diff
@@ -164,6 +164,18 @@ public virtual bool AttachCarryable(Actor self, Actor carryable)
 			CarryableOffset = OffsetForCarryable(self, carryable);
 			aircraft.LandAltitude = aircraft.LandAltitude - new WDist(CarryableOffset.Z);
+
+			aircraft.GetPosition = () =>
```
Our usual strategy for cases like this is to define a new interface (maybe `IAircraftCenterOffset` and `IPreventAircraftLanding`?) and have `Aircraft` ask the trait(s) for their values when needed.

This helps decouple the traits, improves moddability, and for things like offsets makes it easy to aggregate over multiple active traits.

tovl replied:
Done. Although it should be noted that we don't just prevent landing in some terrain, we also need to enable landing in other terrain (amphibious vehicles for instance).

pchote commented on OpenRA.Mods.Common/Activities/UnloadCargo.cs:
```diff
@@ -24,13 +24,18 @@ public class UnloadCargo : Activity
+		bool moved;
 
-		public UnloadCargo(Actor self, bool unloadAll)
+		public UnloadCargo(Actor self, Target? destination = null, bool unloadAll = true)
```
Likewise here wrt the nullable target.

tovl replied:
Fixed.

pchote commented on OpenRA.Mods.Common/Scripting/Properties/TransportProperties.cs:
```diff
@@ -42,9 +42,9 @@ public TransportProperties(ScriptContext context, Actor self)
 		[ScriptActorPropertyActivity]
 		[Desc("Command transport to unload passengers.")]
-		public void UnloadPassengers()
+		public void UnloadPassengers(Target? destination = null)
```
This one needs to stay nullable because of the way Lua and our custom bridge works.

tovl replied:
Fixed.

pchote commented:
Dependencies have been merged.

tovl commented on OpenRA.Mods.Common/Activities/Air/Land.cs:
```diff
@@ -17,25 +18,37 @@ namespace OpenRA.Mods.Common.Activities
+		bool finishedApproach;
 
-		public Land(Actor self, Target t, WVec offset)
+		public Land(Actor self, Target? target = null, WVec? offset = null, int? facing = null)
```
The problem with that is that the nullability of  `target` and `facing` does carry meaning. 

`target` needs to default to `null` because we have to distinguish between an invalid target and no target at all. In fact, I made a mistake here: When `target == null` the destination should be assigned on `OnFirstRun`, not in the constructor. Otherwise this wont work with queued orders. If we want to polish the landing location for fixed wing aircraft further in the future, then target would probably need to be assigned even later in the activity. If we assign `Target.Invalid` as a placeholder, we can no longer distinguish between the case when we should land in place and the case when we want to land on an actor that is dead or otherwise has become invalid.

`facing` needs to default to `null` as well, because when no facing is given we should not assume any magic value for facing. `aircraft.Info.InitialFacing` is only a valid default when `aircraft.Info.TurnToLand` is true, which isn't the case for any unit in the official mods as far as I'm aware. In the regular case a null facing should always stay null and be referenced throughout the `Land` activity to indicate no special turns need to be made.

I can work around this, but doing so would mean that any activity that calls `Land` as a childactivity and optionally passes a destination and/or facing (which is most of them now) would need to call `Land` multiple times in an if/else statement depending on whether the destination in the parent activity is null or not, etc. Any new activity which does so would need to have this as well leading to potential for errors and code bloat. Whereas if these parameters are nullable they can just be passed along directly (through multiple layers of parent/child activities if necessary) without regard for whether they are null or not.

Once again, I can work around these limitations if necessary (I'm working on it now), but it just doesn't sit well with me. If the idea is to not expose implementation details in the public interface, it seems that in a practical sense this would actually be doing the opposite.

If wrapping arguments in `Nullable<>` is really objectionable, we should probably look into adding support for NA values in our custom structs instead to handle cases such as these (there are probably many more cases where such a thing would come in handy). 

Thinking some more about it,  why is `Target` a `struct` in the first place and not a `class`?

pchote replied:
> If we assign `Target.Invalid` as a placeholder, we can no longer distinguish between the case when we should land in place and the case when we want to land on an actor that is dead or otherwise has become invalid.

Right. In this case, it makes sense to keep the target nullable internally, but then exposed via a constructor that takes a `Target` (which is assigned), and one that doesn't (which assigns `null`).

> `facing` needs to default to `null` as well, because when no facing is given we should not assume any magic value for facing. 

Our standard convention is to pass `facing: -1` in these cases.

> Thinking some more about it,  why is `Target` a `struct` in the first place and not a `class`?

Performance. Structs live on the stack or directly in the class memory on the heap, so avoid a level of pointer indirection to access.

tovl replied:
Replaced all `Nullable<>` wrappers in constructors. I also realized we already have NA values for 'Target' in a way, we just don't have a way to check for them without lumping in unrelated stuff like dead actors, so I added a simple check for that.

tovl commented:
Rebased.

tovl commented:
> Take-off / Landing conflict

This would be much easier to fix after the `TakeOff` refactor in #16483. So to avoid duplication, I'd rather fix that in a separate PR after both this and #16483 are merged.

matjaeck commented:
> Only allow transport / carryalls to force land

@pchote noted that the intention is to allow force landing for all helicopters, so this has been settled.

tovl commented:
> RTB polish

This is fixed now.

> stop order can prevent take off

This should be fixed now.

> I find it weird that they can stay landed anywhere but on the helipad.

It is indeed inconsistent, so I've made force-landing work on helipads as well. This suppresses the take-off on resupply behaviour. Also, helicopters will no longer try to land on helipads if they would immediately take off again (unless forced to land).

pchote commented:
> It is indeed inconsistent, so I've made force-landing work on helipads as well. This suppresses the take-off on resupply behaviour. 

Nice. This provides a proper fix for #5709.

reaperrr commented:
D2k Frigates (and therefore TD C17, though I didn't test) still can't land.

The problem here is that `AircraftCanEnter` is currently only able to return `true` for aircraft that try to enter an actor for resupply, so these `ProductionAirdrop` actors currently still need `LandableTerrainTypes` to be able to land on the airfield/starport.

pchote commented:
Considering that both actors are purely visual and don't/shouldn't interact with other actors in the world I suggested in #16574 to add an `OccupiesSpace` flag that can turn off the checks and cell reservations.

tovl commented:
Airdrop landing has been fixed.

pchote commented:
The issue with transport planes doing loops still persists - enable cheats and build several vehicles from a Nod airstrip in TD for a simple repro case.

tovl commented:
I know. I'm not sure I'd call that a problem. It makes sense for the planes to keep their distance. Also, build times should never be so short as for this to be relevant anyway.

pchote commented:
> I know. I'm not sure I'd call that a problem. It makes sense for the planes to keep their distance. Also, build times should never be so short as for this to be relevant anyway.

My issue here is that for all intents and purposes the c17/frigate are animations on the airfield/starport - they aren't targetable/killable, don't repulse, and don't provide vision.  Having them circle before landing feels inconsistent with this, so without there being any gameplay improvements to trade this off its going to be really frustrating (for me at least) to deal with this when debugging with the instant build cheat enabled.

However, there *is* a clear gameplay improvement that can be solved to make this wortwhile: if the airfield exit is blocked (which can happen in normal gameplay) units will still be delivered, but all stack in the same cell. If we could make the aircraft check the structure's production exit cell (and reserve it so that other ground units cant block it after landing has started) then we have an easy fix for this long standing issue.

A future PR could then polish this further by introducing a proper holding pattern activity to make the aircraft fly laps around the airfield until the exit is clear.

tovl commented:
Added a check for blocked exit cells for airlifts.

tovl committed:
Prevent infinite loop between FlyAttack and ReturnToBase.

tovl commented:
Added a bugfix for the issue with selling the airfield and planes hanging still in the air.

> Helicopters will stay landed on the ground, but they take off after selling the helipad if they're ordered with RTB or landing order to resupply.

I'm not sure what the bug is here. What are they supposed to do instead?

matjaeck commented:
> I'm not sure what the bug is here. What are they supposed to do instead?

They should take off, like they they do when ordered with RTB or land order. See #11620 / #15060. not that this, as mentioned, only happens when they're automatically approaching the helipad (force fire somewhere without ammo).

tovl commented:
Ah, I thought you said the problem was that they _did_ take off. This got me confused. :smile: 

Anyway, the same bug-fix as for non-vtol aircraft should have fixed this as well.

tovl commented:
Fixed.

matjaeck commented:
LGTM after @pchote's requests in https://github.com/OpenRA/OpenRA/pull/16509#issuecomment-495948628 are addressed. The td c17 should only be able to fly straightforward. FWIW a plane may overlap with other planes on landing if the build times are ridiculously short.

pchote commented:
#16584 should hopefully be merged this weekend, so we'll be able to switch attention to this again soon.

pchote commented on OpenRA.Mods.Common/Activities/Air/Land.cs:
```diff
@@ -17,25 +19,40 @@ namespace OpenRA.Mods.Common.Activities
+		bool finishedApproach;
 
-		public Land(Actor self, Target t, WVec offset)
+		public Land(Actor self, Target target = new Target(), int facing = -1, WVec offset = new WVec(), CPos[] clearCells = null)
```
I know that my complaining over this must be getting really annoying now, but my motivation comes from dealing with an endless series of bugs like #15520 and #16473 caused by bad/missing handling of default values (`CPos.Zero`), and from all the regressions caused by #15866 and the massive scope creep required to fix them because none of us appreciated all the subtleties of target handling before/during review.

Something like 5a531cf913 or 61dfaa1aba may not be particularly elegant, but sticking to the same boring code patterns used everywhere else helps reduce future "WTF?!?"s and unexpected regressions the next time this code needs to be refactored.

Of course the real issue here is that `Target` is no longer really fit for purpose, and needs a general overhaul to properly represent our newer usecases:
* #16270
* Representing actors that we last saw at position X but are not currently visible
* Handling actors that change state - frozen/visible, transforms, ownership, etc
* Representing "no target" as distinct from "target which is no longer valid" - and making "no target" the default value

This is a major project in its own right, and IMO we should stay well clear of it until after the activity rework is finished and stable.

tovl replied:
I'm a bit confused. Introducing `Target.IsNone` is supposed to alleviate these very issues. Specifically:

> Representing "no target" as distinct from "target which is no longer valid" - and making "no target" the default value

is exactly what I have done here. So I'm not sure what you're asking. Is there anything in particular that is wrong with my implementation? Or do you mean to say that the general idea is right, but should be handled in a separate PR? Or do I misunderstand completely?

pchote replied:
Sorry for the confusion. What I meant was that the general idea is good, but that we really need to implement this as a first-class `TargetType` and use it everywhere that it is relevant, instead of bolting it on as a workaround for just these specific activities.

tovl replied:
Okay, that makes sense. My idea was that it could be introduced more gradually. Because it doesn't change anything for use-cases that do not actually call the introduced method, it is guaranteed to not cause regressions in code that I didn't touch. It could then be used in other places where it makes sense at a later stage. However, you may be right that reworking `Target` in a more holistic fashion is better.

tovl replied:
Reverted the addition of `Target.IsNone` and reworked based on the pattern of 5a531cf

pchote commented on OpenRA.Mods.Common/Orders/EnterAlliedActorTargeter.cs:
```diff
@@ -16,11 +16,11 @@ namespace OpenRA.Mods.Common.Orders
 
 		public EnterAlliedActorTargeter(string order, int priority,
-			Func<Actor, bool> canTarget, Func<Actor, bool> useEnterCursor)
+			Func<Actor, TargetModifiers, bool> canTarget, Func<Actor, bool> useEnterCursor)
```
#16607 solves the same use case by adding a `Func<bool> requireForceMove` argument.
I like that this approach doesn't hardcode a new concept with the targeter, but it has the downside of requiring shims in `Repairable` to translate.

Do you have a strong preference or thoughts on either of the two approaches?

tovl replied:
Like you said, either approach has downsides and upsides so I don't have strong preference either way. That said, an added benefit of this approach is that it works equally for any modifier key, not just `ForceMove`. That makes this a really generic approach that could potentially be repurposed for many different concepts. Considering this, I have a slight preference for this approach.

pchote replied:
Changed #16607 to use this approach. If we merge that first it will reduce the diff here a bit.

pchote commented on OpenRA.Mods.Common/Orders/EnterTransportsTargeter.cs:
```diff
@@ -20,7 +20,7 @@ public class EnterTransportsTargeter : EnterAlliedActorTargeter<CargoInfo>
 		readonly AlternateTransportsMode mode;
 
 		public EnterTransportsTargeter(string order, int priority,
```
The current version of #16607 also removes this completely.

tovl replied:
Removed in rebase.

pchote commented on OpenRA.Mods.Common/Scripting/Properties/TransportProperties.cs:
```diff
@@ -42,9 +42,9 @@ public TransportProperties(ScriptContext context, Actor self)
 		[ScriptActorPropertyActivity]
 		[Desc("Command transport to unload passengers.")]
-		public void UnloadPassengers()
+		public void UnloadPassengers(Target destination = new Target())
```
I missed this in my examples above.

pchote commented on OpenRA.Mods.Common/Traits/Air/Aircraft.cs:
```diff
@@ -880,16 +921,19 @@ public IEnumerable<IOrderTargeter> Orders
 		{
 			get
 			{
+				yield return new EnterAlliedActorTargeter<BuildingInfo>("ForceEnter", 6,
```
It would in principle be cleaner if `ForceEnter` was implemented via a flag in the `ExtraData` of the `EnterOrder`, and likewise for `Land` on `Move`. I'm not sure how to cleanly implement the order targeter side of this, so this consider this a wishful thought rather than a request.

tovl replied:
I agree, but I don't know how to do this either so I'll leave it like this.

pchote commented on OpenRA.Mods.Common/Traits/Cargo.cs:
```diff
@@ -249,11 +258,43 @@ internal void UnreserveSpace(Actor a)
+
+			state = State.Locked;
+
+			self.CancelActivity();
```
Do we really need to cancel the activity queue here?

The current/previous code paused the transport while loading, but allowed it to continue moving and doing any other queued activities afterwards.

tovl replied:
The `Land` and `WaitFor` activities need to be inserted at the front of the queue interrupting any current activity if possible. I don't know how we can continue the previous queue where it left of without running into trouble with activity states. I'm not sure if `ChildActivities` are any help here.

pchote replied:
The current/old approach for locking the actor while loading was to use `LoadingCondition` to pause `Mobile`. I had always assumed that we would hook that up to `Aircraft`'s `ForceLandCondition` - this was one of the motivations for adding that in the first place.

tovl replied:
`ForceLandCondition` as currently implemented also cancels the activity queue so that solves nothing. I would prefer to leave this issue for a future follow-up.

pchote replied:
Leaving this to a followup is reasonable, but I do think the `ForceLandCondition` is the way to go - theres no conceptual need for that to be cancelling activities even though it is doing that now.

pchote commented on OpenRA.Mods.Common/TraitsInterfaces.cs:
```diff
@@ -437,6 +437,16 @@ public interface IMove
 		bool CanEnterTargetNow(Actor self, Target target);
 	}
 
+	public interface IPositionOffset
```
Can we please call this `IAircaftCenterPositionOffset`, since this only applies to aircraft.

tovl replied:
I deliberately gave it a more generic name since the concept doesn't have to be exclusive to aircraft. It includes horizontal offsets as well, which could be useful for extending the carryall logic to land units.

pchote replied:
My issue here is that right now it *is* exclusive to aircraft, so it is misleading to name it otherwise.
We can rename it to `ICenterPositionOffset` in the future if `Mobile`or other traits add support.

tovl replied:
Changed the name.

tovl commented on OpenRA.Mods.Common/Activities/Air/Land.cs:
```diff
@@ -17,25 +19,40 @@ namespace OpenRA.Mods.Common.Activities
+
+			// NOTE: desiredFacing = -1 means we should not prefer any particular facing and instead just
+			// use whatever facing gives us the most direct path to the landing site.
+			if (facing == -1 && aircraft.Info.TurnToLand)
```
I had thought about it but decided it was better to have the `TurnToDock` case be handled by `ReturnToBase` instead and have it pass a facing to Land. This keeps the landing code more general and keeps everything that is specific to docking in `ReturnToBase.` I also think that for the future we should consider making the preferred dock facing (if any) a characteristic of the airfield/helipad instead of the aircraft. This would also imply unhardcoding the eastward landing direction for airfields.

tovl commented on OpenRA.Mods.Common/Activities/Air/Land.cs:
```diff
@@ -48,61 +65,173 @@ public override Activity Tick(Actor self)
+				cells.Add(landingCell);
+				var blockingCells = cells.ToArray();
+
+				if (!aircraft.CanLand(blockingCells, target.Actor))
```
Done.

tovl commented on OpenRA.Mods.Common/Activities/Air/Land.cs:
```diff
@@ -48,61 +65,173 @@ public override Activity Tick(Actor self)
+				if (aircraft.Info.LandingSounds.Length > 0)
+					Game.Sound.Play(SoundType.World, aircraft.Info.LandingSounds, self.World, aircraft.CenterPosition);
+
 				aircraft.AddInfluence(landingCell);
```
Done.

tovl commented on OpenRA.Mods.Common/Activities/Air/HeliFly.cs:
```diff
@@ -108,7 +108,8 @@ public override Activity Tick(Actor self)
 
 			// Update facing
-			var delta = checkTarget.CenterPosition - aircraft.CenterPosition;
+			var selfPosition = aircraft.GetPosition();
```
Fixed.

tovl commented on OpenRA.Mods.Common/Activities/Air/Land.cs:
```diff
@@ -48,61 +65,173 @@ public override Activity Tick(Actor self)
 			}
 
+			// Reevaluate target position in case the target has moved.
+			targetPosition = this.target.CenterPosition + this.offset;
```
Fixed.

tovl commented on OpenRA.Mods.Common/Activities/Air/Land.cs:
```diff
@@ -48,61 +65,173 @@ public override Activity Tick(Actor self)
+			landingCell = self.World.Map.CellContaining(targetPosition);
+
+			// We are already at the landing location.
+			if ((targetPosition - aircraft.GetPosition()).Length == 0)
```
Fixed by changing to `LengthSquared`. It definitely should not be `HorizontalLengthSquared` because that would mean the aircraft would be content hovering above the landing spot instead of actually landing.

tovl commented on OpenRA.Mods.Common/Activities/EnterTransport.cs:
```diff
@@ -24,6 +24,7 @@ class EnterTransport : Enter
 
 		Actor enterActor;
 		Cargo enterCargo;
+		Aircraft enterAir;
```
Fixed.

pchote commented:
Needs a rebase now, sorry!

reaperrr commented:
#16607 was merged as well, which might make the rebase a bit worse but should at least avoid another one afterwards.

matjaeck commented:
This crashes now when you build something from the TD airstrip. See log below.

<details>

```
OpenRA engine version {DEV_VERSION}
Tiberian Dawn mod version {DEV_VERSION}
on map 360c76d36c7633ad989553acd493be3ab204894e (Back to Basics by pulse).
Date: 2019-06-08 10:06:46Z
Operating System: Linux (Unix 4.15.0.50)
Runtime Version: Mono 5.20.1.19 (tarball Thu Apr 11 09:02:17 UTC 2019) CLR 4.0.30319.42000
Exception of type `System.ArgumentNullException`: Value cannot be null.
Parameter name: source
  at System.Linq.Enumerable.Count[TSource] (System.Collections.Generic.IEnumerable`1[T] source) [0x0000d] in <35ad2ebb203f4577b22a9d30eca3ec1f>:0 
  at OpenRA.Mods.Common.Traits.Aircraft.OccupiedCells () [0x0000d] in <4f387a97abe440c8b25182d7febee932>:0 
  at OpenRA.Mods.Common.Traits.ActorMap.AddInfluence (OpenRA.Actor self, OpenRA.Traits.IOccupySpace ios) [0x00000] in <4f387a97abe440c8b25182d7febee932>:0 
  at OpenRA.World.AddToMaps (OpenRA.Actor self, OpenRA.Traits.IOccupySpace ios) [0x00000] in <7620344bc8934941b5d580215ab85e84>:0 
  at OpenRA.Mods.Common.Traits.Aircraft.AddedToWorld (OpenRA.Actor self) [0x00000] in <4f387a97abe440c8b25182d7febee932>:0 
  at OpenRA.Mods.Common.Traits.Aircraft.OpenRA.Traits.INotifyAddedToWorld.AddedToWorld (OpenRA.Actor self) [0x00000] in <4f387a97abe440c8b25182d7febee932>:0 
  at OpenRA.World.Add (OpenRA.Actor a) [0x0003a] in <7620344bc8934941b5d580215ab85e84>:0 
  at OpenRA.World.CreateActor (System.Boolean addToWorld, System.String name, OpenRA.Primitives.TypeDictionary initDict) [0x0003c] in <7620344bc8934941b5d580215ab85e84>:0 
  at OpenRA.World.CreateActor (System.String name, OpenRA.Primitives.TypeDictionary initDict) [0x00000] in <7620344bc8934941b5d580215ab85e84>:0 
  at OpenRA.Mods.Cnc.Traits.ProductionAirdrop+<>c__DisplayClass1_0.<Produce>b__0 (OpenRA.World w) [0x00085] in <4810461fccf94b0287512368d81ed1ca>:0 
  at OpenRA.World.Tick () [0x001a9] in <7620344bc8934941b5d580215ab85e84>:0 
  at OpenRA.Game.InnerLogicTick (OpenRA.Network.OrderManager orderManager) [0x001bc] in <7620344bc8934941b5d580215ab85e84>:0 
  at OpenRA.Game.LogicTick () [0x0004a] in <7620344bc8934941b5d580215ab85e84>:0 
  at OpenRA.Game.Loop () [0x000e3] in <7620344bc8934941b5d580215ab85e84>:0 
  at OpenRA.Game.Run () [0x0003c] in <7620344bc8934941b5d580215ab85e84>:0 
  at OpenRA.Game.InitializeAndRun (System.String[] args) [0x00010] in <7620344bc8934941b5d580215ab85e84>:0 
  at OpenRA.Program.Main (System.String[] args) [0x00044] in <7620344bc8934941b5d580215ab85e84>:0

tovl commented:
Fixed and rebased.

pchote commented on OpenRA.Mods.Common/Traits/Repairable.cs:
```diff
@@ -64,7 +64,7 @@ public IEnumerable<IOrderTargeter> Orders
 			get
 			{
-				yield return new EnterAlliedActorTargeter<BuildingInfo>("Repair", 5, CanRepairAt, _ => CanRepair() || CanRearm());
+				yield return new EnterAlliedActorTargeter<BuildingInfo>("Repair", 5, (a, m) => CanRepairAt(a), _ => CanRepair() || CanRearm());
```
This change should be dropped to avoid regressing #16607

tovl replied:
Reverted.

pchote commented on mods/ts/rules/defaults.yaml:
```diff
@@ -864,6 +864,8 @@
 		Voice: Move
 		IdealSeparation: 853
 		MaximumPitch: 120
+		LandableTerrainTypes: Clear,Road,Rail,DirtRoad,Rough,Tiberium,BlueTiberium,Veins
```
Please add spaces between the items.

tovl replied:
Done.

matjaeck commented:
This should fix #14102 (same fix as https://github.com/OpenRA/OpenRA/pull/16509#issuecomment-496015818).

matjaeck commented:
This should fix #16646 (same fix as https://github.com/OpenRA/OpenRA/pull/16509#issuecomment-496015818).

tovl commented on OpenRA.Mods.Common/Activities/Air/Land.cs:
```diff
@@ -48,76 +65,186 @@ public override Activity Tick(Actor self)
 			}
 
+			// Move towards landing location
+			var ownPosition = aircraft.GetPosition();
```
Fixed.

tovl commented on OpenRA.Mods.Common/Activities/Air/Land.cs:
```diff
@@ -48,76 +65,186 @@ public override Activity Tick(Actor self)
-			}
-
-			// For VTOLs we assume we've already arrived at the target location and just need to move downward
+			// Final descend.
```
Fixed.

tovl commented on OpenRA.Mods.Common/Traits/Air/Aircraft.cs:
```diff
@@ -236,6 +237,15 @@ public Aircraft(ActorInitializer init, AircraftInfo info)
+			return pos;
+		}
+
 		public void AddLandingOffset(int offset)
```
Removed and used `IPositionOffset` instead.

tovl commented on OpenRA.Mods.Common/Activities/Air/Land.cs:
```diff
@@ -48,76 +65,186 @@ public override Activity Tick(Actor self)
+			if (!aircraft.Info.VTOL && !finishedApproach)
+			{
+				// Calculate approach trajectory
+				Calculate(self);
```
Done.

tovl commented on OpenRA.Mods.Common/Activities/Air/ReturnToBase.cs:
```diff
@@ -175,62 +118,39 @@ public override Activity Tick(Actor self)
+					QueueChild(self, new FlyCircle(self, aircraft.Info.NumberOfTicksToVerifyAvailableAirport), true);
+					return this;
 				}
 				else
```
Done.

tovl commented on OpenRA.Mods.Common/Activities/Air/ReturnToBase.cs:
```diff
@@ -175,62 +118,39 @@ public override Activity Tick(Actor self)
 				}
 				else
 				{
 					// Prevent an infinite loop in case we'd return to the activity that called ReturnToBase in the first place. Go idle instead.
```
Yes. In fact it was misleading before, because it didn't actually prevent the loop: see b187571

tovl commented on OpenRA.Mods.Common/Activities/Air/TakeOff.cs:
```diff
@@ -24,15 +24,16 @@ public class TakeOff : Activity
 		bool assignTargetOnFirstRun;
 
-		public TakeOff(Actor self, Target target)
+		public TakeOff(Actor self, Target target, bool interruptible = true)
```
This is most likely redundant now, so removed it.

tovl commented on OpenRA.Mods.Common/Traits/Air/Aircraft.cs:
```diff
@@ -567,29 +573,66 @@ public WVec FlyStep(int speed, int facing)
 			if (otherActor == self || otherActor == ignoreActor)
 				return false;
 
+			if (!checkMobile && self.Owner.Stances[otherActor.Owner] == Stance.Ally &&
```
Renamed to avoid confusion.

tovl commented on OpenRA.Mods.Common/Traits/Air/Aircraft.cs:
```diff
@@ -874,16 +922,19 @@ public IEnumerable<IOrderTargeter> Orders
+
 				yield return new EnterAlliedActorTargeter<BuildingInfo>("Enter", 5,
-					AircraftCanEnter, target => Reservable.IsAvailableFor(target, self));
+					(target, modifers) => AircraftCanEnter(target), target => Reservable.IsAvailableFor(target, self));
```
Fixed.

reaperrr commented:
> Traits\Air\Aircraft.cs(238,29): error SA1500: Braces for multi-line statements should not share line [C:\projects\openra\OpenRA.Mods.Common\OpenRA.Mods.Common.csproj]
Traits\Air\Aircraft.cs(245,6): error SA1500: Braces for multi-line statements should not share line [C:\projects\openra\OpenRA.Mods.Common\OpenRA.Mods.Common.csproj]
Traits\Air\Aircraft.cs(245,4): error SA1500: Braces for multi-line statements should not share line [C:\projects\openra\OpenRA.Mods.Common\OpenRA.Mods.Common.csproj]

matjaeck commented:
Last changes fixed another issue with blocked helipads which is the same as #14102 I think - see below:

<details>

bleed

![rearmnotblocked](https://user-images.githubusercontent.com/37185497/59151564-9fc68c80-8a35-11e9-8c67-4a379f57cc37.gif)


Before update (2nd heli can not land, #14102)

![rearmgblocked](https://user-images.githubusercontent.com/37185497/59151567-ab19b800-8a35-11e9-9fc3-d6095156fb1f.gif)


Current change

![hoveroverusedpad](https://user-images.githubusercontent.com/37185497/59151569-b371f300-8a35-11e9-9fc2-9859d2bbaf4b.gif)

</details>

On the last image you see that the helicopter is waiting to rearm just above a landed helicopter. Could you polish this so that it waits a bit next to the pad?

tovl commented on OpenRA.Mods.Common/Activities/UnloadCargo.cs:
```diff
@@ -62,48 +72,69 @@ public override Activity Tick(Actor self)
+				else if (!destination.IsNone())
+				{
+					var cell = self.World.Map.Clamp(this.self.World.Map.CellContaining(destination.CenterPosition));
+					QueueChild(self, new Move(self, cell, WDist.FromCells(5)), true);
```
Done.

tovl commented on OpenRA.Mods.Common/Scripting/Properties/TransportProperties.cs:
```diff
@@ -42,9 +42,9 @@ public TransportProperties(ScriptContext context, Actor self)
 		[ScriptActorPropertyActivity]
 		[Desc("Command transport to unload passengers.")]
-		public void UnloadPassengers()
+		public void UnloadPassengers(Target destination = default(Target))
```
Fixed.

tovl commented:
> Could you polish this so that it waits a bit next to the pad?

I think that can best be solved together with the other holding pattern polish issues in a separate PR.

pchote commented:
Would you consider it unreasonable to split this into a couple of smaller PRs to help focus review and testing? Perhaps

1. Merging carryall/rtb into land + force-move landing
2. Production airdrop fixes
3. Land-nearby changes/cleanups
4. Airborne transport changes

Also note that i'm travelling for work for the next week-and-a-bit so my review activity will drop significantly.

tovl commented:
> Would you consider it unreasonable to split this into a couple of smaller PRs to help focus review and testing?

To be honest I'm not particularly enthusiastic about trying to untangle all that this late in the game. Also, keep in mind that the things that could be disentangled most readily are all things that where scope-creeped in specifically in response to review requests. I'm afraid that splitting it up would result in the same points being rehashed again and potentially be inviting more scope-creep.

I understand your frustration with reviewing smorgasbord PRs, but I think we should just remember to be more disciplined about it in the future instead of retroactively spliting them up 115 comments in.

I will revert the `Target` changes and after that we should really aim to wrap things up.

pchote commented:
That's what I was expecting, so i'm not too fussed.

reaperrr commented:
> Also, keep in mind that the things that could be disentangled most readily are all things that where scope-creeped in specifically in response to review requests.

I know that feeling...

tovl commented on OpenRA.Mods.Common/Activities/UnloadCargo.cs:
```diff
@@ -62,48 +72,69 @@ public override Activity Tick(Actor self)
 
-			if (!cargo.CanUnload())
+			// Move to the target destination
+			if (!moved)
```
That doesn't quite work out, but moving the block entirely into `OnFirstRun` also gets rid of `moved`.

teinarss commented on OpenRA.Mods.Common/Traits/Air/Aircraft.cs:
```diff
@@ -233,17 +233,27 @@ public Aircraft(ActorInitializer init, AircraftInfo info)
+			get
+			{
+				var alt = Info.LandAltitude;
+				foreach (var offset in self.TraitsImplementing<IPositionOffset>())
```
Wouldnt it make sense to query this in the constructor?

tovl replied:
The value returned by `PositionOffset()` does not remain constant during the lifetime of the actor. In the case of `Carryall.PositionOffset` it changes depending on what unit it is carrying and even the facing at that particular moment.

teinarss replied:
I meant the query for getting the TraitsImplementing<IPositionOffset>() :smile:

tovl replied:
Fixed.

teinarss commented on OpenRA.Mods.Common/Traits/Air/Aircraft.cs:
```diff
@@ -233,17 +233,27 @@ public Aircraft(ActorInitializer init, AircraftInfo info)
 		{
-			LandAltitude -= new WDist(offset);
+			var pos = self.CenterPosition;
+			foreach (var offset in self.TraitsImplementing<IPositionOffset>())
```
Same here

tovl replied:
Fixed.

teinarss reviewed (dismissed):
Looks good to me now and have tested it in TD.

pchote commented on OpenRA.Mods.Common/Traits/Air/Aircraft.cs:
```diff
@@ -923,7 +971,21 @@ public void ResolveOrder(Actor self, Order order)
+			}
+			else if (orderString == "Enter" || orderString == "ForceEnter" || orderString == "Repair")
 			{
 				// Enter and Repair orders are only valid for own/allied actors,
```
Please add ForceEnter here too

tovl replied:
Done.

pchote commented on OpenRA.Mods.Common/Traits/Cargo.cs:
```diff
@@ -47,9 +47,21 @@ public class CargoInfo : ITraitInfo, Requires<IOccupySpaceInfo>
 		public readonly int PassengerFacing = 128;
 
+		[Desc("How long should this actor wait after loading")]
+		public readonly int WaitAfterLoad = 8;
```
Trait naming conventions are to use `*Delay` for delays, and to mention that they are in ticks in the description.

pchote replied:
These could therefore become
```csharp
[Desc("Delay (in ticks) before continuing after loading a passenger.")]
public readonly int AfterLoadDelay = 8;

[Desc("Delay (in ticks) before unloading the first passenger.")]
public readonly int BeforeUnloadDelay = 8;

[Desc("Delay (in ticks) before continuing after unloading a passenger.")]
public readonly int AfterUnloadDelay = 25;
```

tovl replied:
Done.

pchote commented on OpenRA.Mods.Common/Traits/Cargo.cs:
```diff
@@ -47,9 +47,21 @@ public class CargoInfo : ITraitInfo, Requires<IOccupySpaceInfo>
 		[Desc("Voice to play when ordered to unload the passengers.")]
 		public readonly string UnloadVoice = "Action";
 
+		[Desc("Radius within which the actor will search for a suitable unloading spot when the ordered location is blocked")]
```
Descriptions should end with a period.

tovl replied:
Done.

pchote commented on OpenRA.Mods.Common/Traits/Carryall.cs:
```diff
@@ -26,14 +26,14 @@ public class CarryallInfo : ITraitInfo, Requires<BodyOrientationInfo>, Requires<
 
 		[Desc("Radius around the target drop location that are considered if the target tile is blocked.")]
-		public readonly WDist DropRange = WDist.FromCells(5);
+		public readonly int DropRange = 5;
```
Please keep this as `WDist` as mentioned above.

pchote commented on OpenRA.Mods.Common/Traits/Carryall.cs:
```diff
@@ -26,14 +26,14 @@ public class CarryallInfo : ITraitInfo, Requires<BodyOrientationInfo>, Requires<
 		[Desc("Delay on the ground while attaching an actor to the carryall.")]
 		public readonly int LoadingDelay = 0;
```
Can we please rename this to `BeforeLoadDelay` to match `Cargo`?

tovl replied:
Done.

pchote commented on OpenRA.Mods.Common/Traits/Carryall.cs:
```diff
@@ -26,14 +26,14 @@ public class CarryallInfo : ITraitInfo, Requires<BodyOrientationInfo>, Requires<
 
-		[Desc("Delay on the ground while detacting an actor to the carryall.")]
+		[Desc("Delay on the ground while detaching an actor from the carryall.")]
 		public readonly int UnloadingDelay = 0;
```
Can we please call this `BeforeUnloadDelay` to match `Cargo`?

tovl replied:
Done.

reaperrr commented on OpenRA.Mods.Common/Activities/Air/ReturnToBase.cs:
```diff
@@ -175,62 +118,37 @@ public override Activity Tick(Actor self)
+				if (aircraft.Info.TurnToDock)
+					facing = aircraft.Info.InitialFacing;
+				if (!aircraft.Info.VTOL)
+					facing = 196;
```
64x3, so 192.

GraionDilach replied:
This should be exposed to YAML longterm.

tovl replied:
@GraionDilach That was my plan for a follow up, yes.

tovl replied:
Fixed the value.

pchote commented:
If there are no objections by the weekend I will start applying my fixup requests so that we can move forward here.

tovl commented:
I have been away for work during the past two weeks, so I didn't have the time to spend on this. I will be back coming weekend and address the review comments.

tovl commented on OpenRA.Mods.Common/Traits/Air/Aircraft.cs:
```diff
@@ -200,16 +201,16 @@ public class Aircraft : ITick, ISync, IFacing, IPositionable, IMove, IIssueOrder
 			return self.CenterPosition - new WVec(WDist.Zero, WDist.Zero, self.World.Map.DistanceAboveTerrain(self.CenterPosition));
 		}
 
+		public bool AtLandAltitude { get { return self.World.Map.DistanceAboveTerrain(CenterPosition) == LandAltitude; } }
```
I don't have access to my IDE right now, but I think you are right.

tovl replied:
Done.

tovl commented on OpenRA.Mods.Common/Activities/Air/TakeOff.cs:
```diff
@@ -97,6 +97,9 @@ public override Activity Tick(Actor self)
 			if (moveToRallyPoint && NextActivity == null)
 			{
+				if (!aircraft.Info.VTOL && assignTargetOnFirstRun)
+					return new FlyTimed(1, self);
```
The idea was to nudge the aircraft into the air so the idle behaviour could take over. It shouldn't perform the `AttackMoveActivity` below because target defaults to `Self.Location` and that doesn't really make sense for a non-VTOL aircraft. I think the recent changes to `TakeOff` have made this partially redundant, because the block above seems to already do the undirected takeoff. However, this still needs a conditional to skip the `AttackMoveActivity`

pchote replied:
I'm thinking mainly about compatibility with e98dba2d0a21b16d508c0f2b5fb5e5e3a72d537b

tovl replied:
Yes, it was a mistake in that respect. I should have used childactivities for this. But like I said, it has probably been made redundant anyway.

tovl replied:
Fixed now.

tovl commented on OpenRA.Mods.Common/Activities/DeliverUnit.cs:
```diff
@@ -20,45 +20,36 @@ public class DeliverUnit : Activity
+		}
 
-		DeliverUnit(Actor self, Target destination)
+		public DeliverUnit(Actor self, Target destination, int nearEnough = 5)
```
That will get us into the whole run-time constant vis-a-vis optional parameters snarl again. I'm not sure that's a good idea if we want this merged quickly.

tovl commented on OpenRA.Mods.Common/Scripting/Global/ReinforcementsGlobal.cs:
```diff
@@ -151,55 +151,10 @@ public Actor[] Reinforce(Player owner, string[] actorTypes, CPos[] entryPath, in
-
-					transport.QueueActivity(new Wait(15));
-				}
+					transport.QueueActivity(new Land(transport, Target.FromCell(transport.World, entryPath.Last()), aircraft.Info.InitialFacing, nearEnough: dropRange));
```
That would work except for the facing. I can add plumbing to `UnloadCargo` to pass an optional facing to `Land`. However, it does seem a bit inconsistent that cargo aircraft turn before unloading only when scripted and not when ordered. I'm not sure what your view on this is.

pchote replied:
For now maybe just add a comment explaining this, and perhaps open a refactoring issue?

tovl replied:
Comment added.

tovl commented on OpenRA.Mods.Common/Traits/Air/Aircraft.cs:
```diff
@@ -533,10 +532,11 @@ public int MovementSpeed
 			{
-				if (landingCell.HasValue)
-					return new[] { Pair.New(landingCell.Value, SubCell.FullCell) };
+				var pairs = new Pair<CPos, SubCell>[landingCells.Count()];
```
Done.

tovl commented on OpenRA.Mods.Common/Traits/Air/Aircraft.cs:
```diff
@@ -553,29 +553,66 @@ public WVec FlyStep(int speed, int facing)
 			if (otherActor == self || otherActor == ignoreActor)
 				return false;
 
+			if (!blockedByMobile && self.Owner.Stances[otherActor.Owner] == Stance.Ally &&
```
Done.

tovl commented on OpenRA.Mods.Common/Traits/Air/Aircraft.cs:
```diff
@@ -860,16 +902,19 @@ public IEnumerable<IOrderTargeter> Orders
+
 				yield return new EnterAlliedActorTargeter<BuildingInfo>("Enter", 5,
-					AircraftCanEnter, target => Reservable.IsAvailableFor(target, self));
+					(target, modifiers) => AircraftCanEnter(target) && (!requireForceMove || modifiers.HasModifier(TargetModifiers.ForceMove)), target => Reservable.IsAvailableFor(target, self));
```
Done.

reaperrr commented:
> OpenRA.Utility(1,1): Error: ts|rules/aircraft.yaml:219 refers to a trait field `WaitAfterUnload` that does not exist on `Cargo`.
OpenRA.Utility(1,1): Error: ts|rules/aircraft.yaml:248 refers to a trait field `LoadingDelay` that does not exist on `Carryall`.
OpenRA.Utility(1,1): Error: ts|rules/aircraft.yaml:249 refers to a trait field `UnloadingDelay` that does not exist on `Carryall`.

tovl commented:
Added update rule.

reaperrr commented:
Works well in-game, as far as I can tell.

On the code side, the only things that stood out to me were the int/WDist issue @pchote pointed out and that the update rule only applied to Carryall, but not Cargo.
I updated the update rule myself to save some time, so +1 from me once @pchote has added his commit on the int/WDist issue.

tovl commented:
`Cargo` didn't have these parameters before, so there is nothing to update.

reaperrr commented:
> Cargo didn't have these parameters before, so there is nothing to update.

Ah damn, that's the disadvantage of looking at commits individually (overlooked their introduction and only saw their rename in the last commit). Will revert that myself.

pchote commented:
As part of fixing the int/WDist issue I am also merging some of the later fixups into the original commits that introduced them. I will comment with the updated branch once I am happy with the result.

tovl committed:
Allow VTOLs to land with force-move

tovl committed:
Overhaul Land activity:

- Landing on an actor is no longer blocked by the underlying terrain
- Land in a nearby cell if the requested location is blocked
- Internally manages the fixed-wing landing sequence
- ProductionAirdrop transport waits until the exit is free before landing

tovl committed:
Allow forced landing on helipads and enforce takeoff otherwise.

tovl committed:
Airborne transports only land to (un)load.

tovl committed:
Improve Carryall behaviour and integration with Aircraft.

pchote committed:
Disable water trails for carried Hover MLRS.

pchote commented:
I've massaged this into: https://github.com/OpenRA/OpenRA/compare/bleed...pchote:orderlanding?expand=1

The main changes were reworking the `nearEnough` code and moving/merging commits so that only the final changes are implemented. There are some other tweaks such as changing `GetLandableTerrain` to actually return terrain (it was a bool) but these should not significantly change the behaviour here.

tovl commented:
Took over pchote's changes without further adaptation.
